lua/plenary/filetype.lua

```diff
@@ -0,0 +1,121 @@
+local Path = require('plenary.path')
+
+local os_sep = Path.path.sep
+
+local filetype = {}
+
+local filetype_table = {
+  extension = {},
+  file_name = {},
+}
+
+filetype.add_table = function(new_filetypes)
+  local valid_keys = {'extension', 'file_name'}
+  local new_keys = {}
+  for k, _ in pairs(new_filetypes) do
+    new_keys[k] = true
+  end
+  for _, k in ipairs(valid_keys) do
+    new_keys[k] = nil
+  end
+
+  for k, v in pairs(new_keys) do
+    error(debug.traceback("Invalid key / value:" .. tostring(k) .. " / " .. tostring(v)))
+  end
+
+
+  if new_filetypes.extension then
+    filetype_table.extension = vim.tbl_extend("force", filetype_table.extension, new_filetypes.extension)
+  end
+
+  if new_filetypes.file_name then
+    filetype_table.file_name = vim.tbl_extend("force", filetype_table.file_name, new_filetypes.file_name)
+  end
+end
+
+filetype.add_file = function(filename)
+  local filetype_files = vim.api.nvim_get_runtime_file(
+    string.format(
+      "data/plenary/filetypes/%s.lua", filename
+    ), true
+  )
+
+  for _, file in ipairs(filetype_files) do
+    local ok, msg = pcall(filetype.add_table, dofile(file))
+    if not ok then
+      error("Unable to add file " .. file .. ":\n" .. msg)
+    end
+  end
+end
+
+filetype._get_extension = function(filename)
+  local ext_with_period = filename:match("^.+(%..+)$")
+  if ext_with_period then
+    return ext_with_period:sub(2)
+  end
+end
+
+filetype._parse_modeline = function(tail)
+  if tail:find('vim:') then
+    return tail:match('.*:ft=([^: ]*):.*$') or ''
+  end
+  return ''
+end
+
+filetype._parse_shebang = function(head)
+  if head:sub(1, 2) == '#!' then
+    return head:match('^.+/e?n?v? ?([^/ %d]*).*$') or ''
+  end
+  return ''
+end
```

Comment:
Not sure how we parse shebang. Maybe we want another table for that. Right now i just return python from `#!/bin/python` or `#!/bin/env python` which works for that but what about `#!/bin/node`. I don't really know what people commonly use here. I only use `python` and `sh`.